File: week1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{
    const int AE = 1;
    const int MC = 2;
    const int Visa = 3;
    
    int brand = 0;
    long card = 0;
    int count = 0;
    
    do
    {
         card = get_long("Number: ");
    }
    while (card <= 0);
    
    for (long temp = card; temp > 0; count++)
    {
        temp /= 10;
    }
            

    
    
    //Luhn's algorithm
    long temp = card;
    int sumLast = 0;
    int sumSec2Last = 0;
    int sum = 0;
    
    int tempX = 0;
    
    while (temp > 0)
    {
        sumLast += temp % 10;
        temp /= 10;
        
        if (temp > 0)
        {
            tempX = temp % 10 * 2;
            if ( tempX < 10)
            {
                sumSec2Last += tempX;
            }
            else
            {
                sumSec2Last += tempX % 10 + tempX / 10;
            }
            
            temp /= 10;
        }
    }
    
    sum = sumLast + sumSec2Last;

    
    if (sum%10 != 0)
    {
        printf("INVALID\n");
    }
    else if (count!=13 && count!=15 && count!=16)
    {
        printf("INVALID\n");
    }
    else
    {

        if (count == 13)
        {
            if (card/(long)pow(10, 12) == 4)
            {
                brand = Visa;
            }
        }
        else if (count == 15)
        {
            if (card / (long)pow(10, 13) == 34 || card / (long)pow(10, 13) == 37)
            {
                brand = AE;
            }
        }
        else
        {
            if (card / (long)pow(10, 14) >= 51 && card / (long)pow(10, 14) <=55)
            {
                brand = MC;
            }
            
            if (card / (long)pow(10, 15) == 4 )
            {
                brand = Visa;
            }
            
        }
        
        if (brand == AE)
        {
        printf("AMEX\n");
        }
        else if (brand == MC)
        {
            printf("MASTERCARD\n");    
        }
        else if (brand == Visa)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    
    
    
    
    
}

File: week6/pset6/credit/credit.py
from cs50 import get_string

def checkBrand(number):
    
    length = len(number)
    
    brand = "INVALID"
    
    if length == 13:
        if number[0] == "4":
            brand = "VISA"

    elif length == 15:
        if number[:2] == "34" or number[:2] == "37":
            brand = "AMEX"

    elif length == 16:
        if number[0] == "4":
            brand = "VISA"
        elif number[:2] == "51" or number[:2] == "52" or number[:2] == "53" or number[:2] == "54" or number[:2] == "55":
            brand = "MASTERCARD"

    return brand


def checkLuhn(number):
    number = int(number)
    
    #sum of every other from second-to-last digit, multplied by 2
    sum1 = 0
    #sum of the rest
    sum2 = 0
    
    while number > 0:
        sum2 += number % 10
        
        number = number // 10
        
        if number > 0:
            tmp = number % 10 * 2
            if tmp < 10:
                sum1 += tmp
            else:
                tmp1 = tmp % 10
                tmp2 = tmp // 10
                sum1 += tmp1 + tmp2
            number = number // 10
        
    total = sum1 + sum2
    if total % 10 == 0:
        return True
    else:
        return False
        
        
def main():
    number = get_string("Number: ")
    
    brand = checkBrand(number)
    if brand != "INVALID":
        flag = checkLuhn(number)
        if flag:
            print(brand)
        else:
            print("INVALID")
    
    else:
        print(brand)

main()
